File: packages/coterie-py/src/coterie/nodes/auditor.py
"""Auditor + Judge nodes for `adversarial` mode.

Pattern: Implementer writes code; Auditor critiques it with a structured findings
JSON; Judge sustains or rejects findings; if any sustained and rounds remain,
loop back to Implementer. Otherwise accept.

The Auditor is just a CLI adapter invocation with a specific prompt — that lives
in the `adversarial` mode module. This module holds:
- the auditor's prompt template (used by mode wiring)
- parsing of findings JSON
- the adversarial Judge node
"""

import json
from typing import Any

from coterie.core.llm.base import LLMClient
from coterie.core.state import CoterieState

AUDITOR_PROMPT_TEMPLATE = """You are an adversarial code auditor. The implementer below produced this work for the subtask:

# Subtask
{subtask}

# Implementer's output
{implementer_output}

Find every plausible defect: bugs, edge cases missed, perf problems, security issues,
missed requirements, unclear code, undocumented invariants. Be aggressive but precise.

Return ONLY a JSON array. No prose, no markdown. Each finding:
{{
  "category": "bug|perf|security|clarity|missed-requirement|edge-case",
  "severity": "low|medium|high|critical",
  "description": "<one sentence>",
  "line_ranges": ["path:start-end", ...]
}}

If you find no defects, return [].
"""


JUDGE_SYSTEM = """You are an impartial judge in an adversarial code review.
You see an implementation and an auditor's findings. Decide which findings are
sustained (real defects the implementer should fix) versus rejected (nitpicks,
false positives, or out of scope).

Sustain a finding when: (1) it identifies a real defect, (2) its severity meets or
exceeds the configured sustain_threshold, AND (3) it is in scope for the subtask.

Return strict JSON only — no prose, no markdown:
{
  "sustained": [<finding_indices>],
  "rejected": [<finding_indices>],
  "verdict": "accept" | "revise",
  "reason": "<2-3 sentences>"
}

Verdict "accept" means the implementation is acceptable (sustained findings, if any,
are minor enough to not require another revision pass). "revise" means the implementer
should take another pass to address the sustained findings.
"""

_SEVERITY_RANK = {"low": 1, "medium": 2, "high": 3, "critical": 4}


def parse_findings_from_output(stdout: str) -> list[dict]:
    """Strip code fences and parse the auditor CLI's stdout as a JSON array of findings."""
    cleaned = stdout.strip()
    if cleaned.startswith("```"):
        lines = cleaned.split("\n")
        cleaned = "\n".join(lines[1:-1] if lines[-1].startswith("```") else lines[1:])
    try:
        data = json.loads(cleaned)
        return data if isinstance(data, list) else []
    except json.JSONDecodeError:
        return []


def make_adversarial_judge_node(llm: LLMClient | None = None):
    def judge(state: CoterieState) -> dict[str, Any]:
        cfg = state["config"]
        adv = cfg["adversarial"]
        judge_cfg = adv.get("judge") or {}
        threshold = judge_cfg.get("sustain_threshold", "medium")
        min_rank = _SEVERITY_RANK.get(threshold, 2)
        max_rounds = adv.get("max_rounds", 3)

        mode_state = dict(state.get("mode_state") or {})
        findings: list[dict] = mode_state.get("auditor_findings", [])
        impl_output = mode_state.get("implementer_output", "")
        round_idx = mode_state.get("round_idx", 0)

        eligible_indices = [
            i
            for i, f in enumerate(findings)
            if _SEVERITY_RANK.get(f.get("severity", "low"), 1) >= min_rank
        ]

        if not eligible_indices:
            mode_state["sustained_findings"] = []
            mode_state["verdict"] = "accept"
            return {
                "mode_state": mode_state,
                "judge_history": [
                    {
                        "step": state.get("current_step_idx", 0),
                        "winner": "implementer",
                        "reason": "no findings met severity threshold",
                        "scores": {"sustained_count": 0},
                    }
                ],
                "status": "done",
            }

        if llm is None:
            # No LLM available — conservatively sustain all eligible findings,
            # but if we're out of rounds we still finish.
            sustained = [findings[i] for i in eligible_indices]
            verdict = "revise"
            reason = "no judge LLM configured; conservatively sustaining all eligible findings"
        else:
            findings_block = "\n".join(
                f"[{i}] severity={findings[i].get('severity')} "
                f"category={findings[i].get('category')}: {findings[i].get('description')}"
                for i in eligible_indices
            )
            prompt = (
                f"Subtask: {state['task']}\n\n"
                f"Implementation output:\n{impl_output[:2000]}\n\n"
                f"Eligible findings (severity >= {threshold}):\n{findings_block}\n\n"
                f"Round: {round_idx + 1} of {max_rounds}"
            )
            raw = llm.chat(JUDGE_SYSTEM, [{"role": "user", "content": prompt}])
            try:
                decision = json.loads(raw)
                sustained = [findings[i] for i in decision.get("sustained", []) if i < len(findings)]
                verdict = decision.get("verdict", "revise")
                reason = decision.get("reason", "")
            except (json.JSONDecodeError, IndexError, TypeError):
                sustained = [findings[i] for i in eligible_indices]
                verdict = "revise"
                reason = f"judge LLM output unparseable: {raw[:120]!r}"

        mode_state["sustained_findings"] = sustained
        mode_state["verdict"] = verdict
        mode_state["round_idx"] = round_idx + 1

        out_of_rounds = mode_state["round_idx"] >= max_rounds
        finished = verdict == "accept" or out_of_rounds or not sustained
        next_status = "done" if finished else "executing"

        return {
            "mode_state": mode_state,
            "judge_history": [
                {
                    "step": state.get("current_step_idx", 0),
                    "winner": "implementer" if verdict == "accept" else "auditor",
                    "reason": reason,
                    "scores": {"sustained_count": len(sustained)},
                }
            ],
            "status": next_status,
        }

    return judge


def adversarial_implementer_prompt(state: CoterieState) -> str:
    """Prompt for the implementer agent. Round 1 = the raw task; rounds 2+ include sustained findings."""
    mode_state = state.get("mode_state") or {}
    round_idx = mode_state.get("round_idx", 0)
    if round_idx == 0:
        return state["task"]
    sustained = mode_state.get("sustained_findings", [])
    critiques = "\n".join(
        f"- [{f.get('severity')}] {f.get('description')}" for f in sustained
    )
    return (
        f"{state['task']}\n\n"
        f"Previous round had sustained critiques. Please address each:\n{critiques}"
    )


def adversarial_auditor_prompt(state: CoterieState) -> str:
    mode_state = state.get("mode_state") or {}
    impl_output = mode_state.get("implementer_output", "")
    return AUDITOR_PROMPT_TEMPLATE.format(
        subtask=state["task"], implementer_output=impl_output[:4000]
    )


def make_record_implementer_output_node():
    """Side-effect node: copy the most recent implementer run.stdout into mode_state."""

    def record(state: CoterieState) -> dict[str, Any]:
        runs = state.get("runs") or []
        last_impl = next(
            (r for r in reversed(runs) if r.get("role") == "implementer"), None
        )
        mode_state = dict(state.get("mode_state") or {})
        if last_impl:
            mode_state["implementer_output"] = last_impl["stdout"]
        return {"mode_state": mode_state, "status": "auditing"}

    return record


def make_record_auditor_findings_node():
    """Side-effect node: parse the most recent auditor run.stdout into mode_state.auditor_findings."""

    def record(state: CoterieState) -> dict[str, Any]:
        runs = state.get("runs") or []
        last_auditor = next(
            (r for r in reversed(runs) if r.get("role") == "auditor"), None
        )
        mode_state = dict(state.get("mode_state") or {})
        if last_auditor:
            mode_state["auditor_findings"] = parse_findings_from_output(last_auditor["stdout"])
        return {"mode_state": mode_state, "status": "judging"}

    return record

File: packages/coterie-js/src/nodes/auditor.ts
/** The adversarial mode's nodes: prompts for the implementer/auditor seats,
 *  recorder nodes that lift their outputs into mode_state, and the judge that
 *  sustains/rejects findings and decides accept vs another round.
 *  Wired into a graph by modes/adversarial.ts. */

import { parseJsonLoose } from "../core/json.js";
import type { LLMClient } from "../core/llm/base.js";
import { type CoterieState, lastRunByRole } from "../core/state.js";

export const AUDITOR_PROMPT_TEMPLATE = `You are an adversarial code auditor. The implementer below produced this work for the subtask:

# Subtask
{subtask}

# Implementer's output
{implementer_output}

Find every plausible defect: bugs, edge cases missed, perf problems, security issues,
missed requirements, unclear code, undocumented invariants. Be aggressive but precise.

Return ONLY a JSON array. No prose, no markdown. Each finding:
{
  "category": "bug|perf|security|clarity|missed-requirement|edge-case",
  "severity": "low|medium|high|critical",
  "description": "<one sentence>",
  "line_ranges": ["path:start-end", ...]
}

If you find no defects, return [].
`;

const JUDGE_SYSTEM = `You are an impartial judge in an adversarial code review.
You see an implementation and an auditor's findings. Decide which findings are
sustained (real defects the implementer should fix) versus rejected (nitpicks,
false positives, or out of scope).

Sustain a finding when: (1) it identifies a real defect, (2) its severity meets or
exceeds the configured sustain_threshold, AND (3) it is in scope for the subtask.

Return strict JSON only — no prose, no markdown:
{
  "sustained": [<finding_indices>],
  "rejected": [<finding_indices>],
  "verdict": "accept" | "revise",
  "reason": "<2-3 sentences>"
}`;

const SEVERITY_RANK: Record<string, number> = { low: 1, medium: 2, high: 3, critical: 4 };

export function parseFindings(stdout: string): any[] {
  const data = parseJsonLoose(stdout);
  return Array.isArray(data) ? data : [];
}

/** Whether the auditor output actually yielded a JSON array of findings — as
 *  opposed to prose, a crash, or malformed JSON. Lets the judge distinguish a
 *  genuine "no defects" ([]) from "couldn't read the auditor", which must not be
 *  silently treated as a clean bill of health. */
export function findingsAreParseable(stdout: string): boolean {
  return Array.isArray(parseJsonLoose(stdout));
}

export function makeAdversarialJudgeNode(llm: LLMClient | null) {
  return async (state: CoterieState) => {
    const cfg = state.config;
    const adv = cfg.adversarial;
    const judgeCfg = adv.judge ?? {};
    const threshold = judgeCfg.sustain_threshold ?? "medium";
    const minRank = SEVERITY_RANK[threshold] ?? 2;
    const maxRounds = adv.max_rounds ?? 2;

    const modeState = { ...(state.mode_state ?? {}) };
    const findings = (modeState.auditor_findings ?? []) as any[];
    const implOutput = (modeState.implementer_output ?? "") as string;
    const roundIdx = (modeState.round_idx ?? 0) as number;

    const eligibleIndices = findings
      .map((f, i) => [i, f] as const)
      .filter(([, f]) => (SEVERITY_RANK[f.severity ?? "low"] ?? 1) >= minRank)
      .map(([i]) => i);

    if (eligibleIndices.length === 0) {
      modeState.sustained_findings = [];
      modeState.verdict = "accept";
      // If the auditor's output couldn't be parsed (crash / prose / bad JSON),
      // "no eligible findings" doesn't mean "clean" — say so rather than implying
      // a pass. (We still terminate; the failure is surfaced in the trace.)
      const unparsed = modeState.auditor_unparsed === true;
      return {
        mode_state: modeState,
        judge_history: [{
          step: state.current_step_idx ?? 0,
          winner: "implementer",
          reason: unparsed
            ? "auditor produced no parseable findings — accepting, but the audit may be incomplete"
            : "no findings met severity threshold",
          scores: { sustained_count: 0 },
        }],
        status: "done" as const,
      };
    }

    let sustained: any[];
    let verdict: string;
    let reason: string;
    if (!llm) {
      sustained = eligibleIndices.map((i) => findings[i]);
      verdict = "revise";
      reason = "no judge LLM configured; conservatively sustaining all eligible findings";
    } else {
      const findingsBlock = eligibleIndices.map((i) =>
        `[${i}] severity=${findings[i].severity} category=${findings[i].category}: ${findings[i].description}`
      ).join("\n");
      const prompt =
        `Subtask: ${state.task}\n\n` +
        `Implementation output:\n${implOutput.slice(0, 2000)}\n\n` +
        `Eligible findings (severity >= ${threshold}):\n${findingsBlock}\n\n` +
        `Round: ${roundIdx + 1} of ${maxRounds}`;
      const raw = await llm.chat(JUDGE_SYSTEM, [{ role: "user", content: prompt }]);
      const decision = parseJsonLoose(raw);
      if (decision && typeof decision === "object") {
        sustained = (decision.sustained ?? []).map((i: number) => findings[i]).filter(Boolean);
        verdict = decision.verdict ?? "revise";
        reason = decision.reason ?? "";
      } else {
        sustained = eligibleIndices.map((i) => findings[i]);
        verdict = "revise";
        reason = `judge LLM output unparseable: ${raw.slice(0, 120)}`;
      }
    }

    modeState.sustained_findings = sustained;
    modeState.verdict = verdict;
    modeState.round_idx = roundIdx + 1;
    const outOfRounds = modeState.round_idx >= maxRounds;
    const finished = verdict === "accept" || outOfRounds || sustained.length === 0;
    // Surface the honest outcome: hitting the round cap with sustained defects
    // still outstanding is "done with unresolved issues", not a clean accept.
    const unresolved = finished && verdict !== "accept" && sustained.length > 0;
    modeState.unresolved_findings = unresolved ? sustained.length : 0;

    return {
      mode_state: modeState,
      judge_history: [{
        step: state.current_step_idx ?? 0,
        winner: verdict === "accept" ? "implementer" : "auditor",
        reason: unresolved ? `${reason} (round cap reached with ${sustained.length} unresolved)` : reason,
        scores: { sustained_count: sustained.length },
      }],
      status: finished ? ("done" as const) : ("executing" as const),
    };
  };
}

export function adversarialImplementerPrompt(state: CoterieState): string {
  const ms = state.mode_state ?? {};
  const round = ms.round_idx ?? 0;
  if (round === 0) return state.task;
  const sustained = (ms.sustained_findings ?? []) as any[];
  const critiques = sustained.map((f) => `- [${f.severity}] ${f.description}`).join("\n");
  return `${state.task}\n\nPrevious round had sustained critiques. Please address each:\n${critiques}`;
}

export function adversarialAuditorPrompt(state: CoterieState): string {
  const ms = state.mode_state ?? {};
  const implOutput = (ms.implementer_output ?? "") as string;
  // Single-pass substitution so a user task containing the literal
  // "{implementer_output}" can't collide with the second placeholder.
  const subs: Record<string, string> = {
    "{subtask}": state.task,
    "{implementer_output}": implOutput.slice(0, 4000),
  };
  return AUDITOR_PROMPT_TEMPLATE.replace(/\{subtask\}|\{implementer_output\}/g, (m) => subs[m] ?? m);
}

export function makeRecordImplementerOutputNode() {
  return async (state: CoterieState) => {
    const lastImpl = lastRunByRole(state.runs, "implementer");
    const modeState = { ...(state.mode_state ?? {}) };
    if (lastImpl) modeState.implementer_output = lastImpl.stdout;
    return { mode_state: modeState, status: "auditing" as const };
  };
}

export function makeRecordAuditorFindingsNode() {
  return async (state: CoterieState) => {
    const lastAuditor = lastRunByRole(state.runs, "auditor");
    const modeState = { ...(state.mode_state ?? {}) };
    if (lastAuditor) {
      modeState.auditor_findings = parseFindings(lastAuditor.stdout);
      // Distinguish a real "[]" (clean) from a crash / prose / bad JSON, so the
      // judge doesn't read a broken audit as a pass.
      modeState.auditor_unparsed =
        lastAuditor.exit_code !== 0 || !findingsAreParseable(lastAuditor.stdout);
    }
    return { mode_state: modeState, status: "judging" as const };
  };
}
